### DISCUSSION TITLE
Bach, Neoscore, CAC in PureData

### DISCUSSION THREAD

**Discussion by enh3** [charlesneimog/py4pd#85] [General]
Hi Charles,

I've just been a deep dive into options for something like the bach paradigm in Max but for puredata, and I came across your project. Another thing I came across was [this](https://arxiv.org/abs/2208.04877) project using INScore. Your idea seems really promising, and neoscore seems to have very similar goals to INScore, though for the moment it seems that INScore is a more mature project. Just for pure curiosity, have you considered basing your project on INScore but decided that neoscore was a better fit? Or have you not come across INScore?

At any rate, this is a very intriguing project that I will certainly be following. Having access to CAC resources in a real-time environment like the bach library has accomplished opens up so many possibilities for interactive algorithmic composition, as well as pedagogical tools and music games. Bach is great, but it would be beautiful to have this in an open source version. puredata being Max's cousin, it seems like the logical candidate.

Just as one other thought, have you considered whether bach, seeing as it's open source, could be ported to puredata? I know this would be a huge task, and Max and puredata have a lot of differences, but it seems maybe less intimidating than building everything from scratch. Maybe one day we could have either bach, neoscore, or INScore for notation visualisation, and then use something like [this](https://github.com/byulparan/pdlisp) to integrate our beloved common lisp ecosystem projects from the PW legacy. This is all just dreaming, but I'm trying to imagine ways that robust CAC could come to pd. Let me know what you think.

Best,
Kjel

**Comment by charlesneimog** [comment]
Hi Kjel, 

I'm glad you found py4pd-score searching for bach and Pd. 

Yes, neoscore is better for me, mainly because in written in `Python` and it is very easy to install in all `OSs`, also with `py4pd-score` I just want to see Partial Tracking analysis, and someday I will implement the tree rhythm syntax of OpenMusic. So, for now, I am not using the interactive part of Neoscore.  I am not sure if you installed `py4pd-score`, but it is a kind of OpenMusic library and is some kind of library of another object called `py4pd` ([Docs](https://charlesneimog.github.io/py4pd/)). Also, InScore is big, so it is hard to understand (at least for me).

Obs: `py4pd` is something like Bach but it is more extensible, mainly because you can use Python to write your own PureData Objects, something like the OM-Libraries ecosystem. Read a little of the Docs, I believe you will understand, mainly the Python Use section.

---------- 
About the second and third paragraph, bring the CAC to PureData is the exactly aim of `py4pd` (not py4pd-score that is just a Library to use with `py4pd`). 

There is some problems to translate the Bach Library for PureData. First, as you point, it is a big work. Second, because its size,  for me don't worst translate some of the Bach things that I don't use in my compositions (I am more a composer than I programmer). Another big problem with Bach is the `C` language. `C` language is amazing and fast but is very hard to program and debug, it is not the language for composers. 

Let me give you an example, now I am interested to use draws to make music, something like the UPIC of Xenakis. With `py4pd` you can make this very easy (see [py4pd-upic](https://github.com/charlesneimog/py4pd-upic), it has just 300 lines of code, in `C` this would be a BIG work. 

So is that, I also love Bach and miss something like Bach in PureData, `py4pd` will be something like Bach but separated into several Libraries.

EDIT: I saw that you work with OM, Python, and the [om-cluster-rules](https://github.com/enh3/om-cluster-rules), maybe you could  use `py4pd` to bring om-cluster-rules for PureData.

**Reply by enh3** [reply]
Hi, Charles,

Thanks for your detailed reply! This all makes a lot of sense, and I can see why you've chosen to go with neoscore over INScore. I also see that you're developing [pd4web](https://github.com/charlesneimog/pd4web) which I imagine might provide some similar functionality to [inscoreweb](https://github.com/grame-cncm/inscoreweb). Part of the reason I'm looking into all this is that in addition to having something like Bach in a fully open source context, it would be nice to be able to create things like interactive collective compositions / pedagogical tools and have them accessible via the web. Both pd4web and inscoreweb seem promising for something like this. 

As for bringing om-cluster-rules to py4pd, we would first have to get the cluster-engine to py4pd, which is definitely beyond my capacity (time and skill) at the moment. For the long term, I'd love to get involved in small ways where I can. I'm looking forward to seeing how this project develops.

PS I posted in py4pd-score just because my question was specifically about the score elements, but I agree that it makes more sense on the main py4pd page!

**Reply by charlesneimog** [reply]
Yes, if you want interactive things related to scores and some kind of audio processing I believe is best to use/study pd4web (that is under development yet) and this is me promoting my work rsrs. I did not use it as a collective composition tool and I also did not know about inscoreweb, I was using VexFlow. Of course is my work, but seems to be promising mainly because of the evolution of WebAssembly (will be more fast and reliable).

**Reply by enh3** [reply]
Great to know. Well thanks again for taking this on. It's very exciting and I'll be keeping an eye on it as it develops. Hopefully when I have more free time I can delve deeper and make some contributions. Take care!

**Comment by charlesneimog** [comment]
Obs.: @enh3 I will move this for `py4pd`, because it is a great topic to discuss.
